File: typescript/src/TennisGame3.ts
import {TennisGame} from './TennisGame';

const pointTitles: string[] = ['Love', 'Fifteen', 'Thirty', 'Forty'];

export class TennisGame3 implements TennisGame {
    private p2: number = 0;
    private p1: number = 0;
    private p1N: string;
    private p2N: string;

    constructor(p1N: string, p2N: string) {
        this.p1N = p1N;
        this.p2N = p2N;
    }

    getScore(): string {
        if (this.noWinnerYet() && this.isBeforeDeuce()) {
            return this.equality() ? this.equalityScore() : this.normalScore();
        } else if (this.equality()) {
            return 'Deuce';
        } else {
            return this.winOrAdvantageScore() + this.findLeaderName();
        }
    }

    private isBeforeDeuce() {
        return this.p1 + this.p2 < 6;
    }

    private noWinnerYet() {
        return this.p1 < 4 && this.p2 < 4;
    }

    private equality() {
        return this.p1 === this.p2;
    }

    private winOrAdvantageScore() {
        return (this.hasWinner() ? 'Win for ' : 'Advantage ');
    }

    private normalScore() {
        return pointTitles[this.p1] + '-' + pointTitles[this.p2];
    }

    private equalityScore() {
        return pointTitles[this.p1] + '-All';
    }

    private hasWinner() {
        return Math.abs(this.p1 - this.p2) > 1;
    }

    private findLeaderName() {
        return this.p1 > this.p2 ? this.p1N : this.p2N;
    }

    wonPoint(playerName: string): void {
        if (playerName === this.p1N)
            this.p1 += 1;
        else
            this.p2 += 1;

    }
}

File: java/src/main/java/TennisGame3.java

public class TennisGame3 implements TennisGame {

    private int p2Score;
    private int p1Score;
    private String p1Name;
    private String p2Name;
    private String[] pointScores = new String[]{"Love", "Fifteen", "Thirty", "Forty"};

    public TennisGame3(String p1N, String p2N) {
        this.p1Name = p1N;
        this.p2Name = p2N;
    }

    public String getScore() {
        if (isInPlayoff()) {
            if (p1Score == p2Score)
                return "Deuce";
            else if (Math.abs(p1Score - p2Score) == 1)
                return "Advantage " + leadingPlayerName();
            else
                return "Win for " + leadingPlayerName();
        } else {
            if (p1Score == p2Score)
                return pointScore(p1Score) + "-All";
            else
                return pointScore(p1Score) + "-" + pointScore(p2Score);
        }
    }

    private String pointScore(int intScore) {
        return pointScores[intScore];
    }

    private String leadingPlayerName() {
        return p1Score > p2Score ? p1Name : p2Name;
    }

    private boolean isInPlayoff() {
        return p1Score >= 4 || p2Score >= 4 || p1Score + p2Score == 6;
    }

    public void wonPoint(String playerName) {
        if (playerName == "player1")
            this.p1Score += 1;
        else
            this.p2Score += 1;

    }

}
